### PR TITLE
[WIP] Allow overriding minimum supported glibc version.

### PR DESCRIPTION
**Pull request by duckinator**
Add a `minimum-glibc-version` config option to package metadata, in case dist's auto-detection doesn't work.

- [x] Add to `cargo_dist::config::v0::DistMetadata`
- [x] Add it to the `basic_cargo_toml_no_change` test.
- [ ] Determine where it should be converted from a `String` to an actual version.
- [ ] Weave it back to `cargo_dist::platform::native_glibc_version()`.
- [ ] Add test(s) order of preference: override > detected > default.

### CODE REVIEW

File: cargo-dist/src/config/v0.rs [hunk 1]
```diff
@@ -784,6 +791,9 @@ impl DistMetadata {
         if mac_pkg_config.is_none() {
             mac_pkg_config.clone_from(&workspace_config.mac_pkg_config);
         }
+        if minimum_glibc_version.is_none() {
```
[the remark this answers is not in the document]
Comment (ashleygwilliams) [reply cargo-dist/src/config/v0.rs:794]: i am in favor of per package, but open to arguments against.

[the remark this answers is not in the document]
Comment (mistydemeo) [reply cargo-dist/src/config/v0.rs:794]: Given that we're building software on the same machine I'd imagine every package will usually get the same glibc - but I can imagine highly specific environments with weird builds that would result in different versions per-package? So yeah, setting it per-package seems fine.

[the remark this answers is not in the document]
Comment (ashleygwilliams) [reply cargo-dist/src/config/v0.rs:794]: yeah my reasoning here comes from the fact that this is a config to be used sparingly, usually when something is already *known* to be weird or has gone terribly wrong ;)

**Comment by mistydemeo** [comment]
> Figure out what (if anything) I need to do in v0_to_v1.rs -- do I need to use minimum_glibc_version?

We do, yes. This is currently one of the more confusing parts of our config system. Let me explain.

The `v0` config represents the "old" config. The `v1` config represents the newer format that we're working towards. Currently, we're in something of an intermediate state: the `v0` config is still the on-disk representation, as one single, large document. The `v1` config is what we use internally.

The `v0_to_v1.rs` stuff is mapping from what we've loaded to what will be in use by the rest of the app. We still (ideally) plan to migrate the on-disk config from the v0 to v1 formats by the time dist 1.0 comes out. In the meantime, adding new config fields does unfortunately mean adding them in both places, with this mapping.

File: cargo-dist/src/config/v0.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (mistydemeo) [reply cargo-dist/src/config/v0.rs:794]: Yep, that's a very good call.

[the remark this answers is not in the document]
Comment (duckinator) [reply cargo-dist/src/config/v0.rs:794]: This was a deliberate choice, because it seems like there may be a benefit to supporting both.

The thing prompting this is Python tools using [`manylinux` (PEP 600)](https://peps.python.org/pep-0600/) tags. E.g., `manylinux2014` would be glibc 2.17+, which is older than cargo-dist's default.

If a workspace has multiple Python package with native extensions and they all target (e.g.) `manylinux2014`, the glibc version would be the same, so it makes sense to be able to specify it at the workspace level.

However, if someone wants to migrate those same packages to `manylinux_2_31_x86_64` (glibc 2.31, cargo-dist default) supporting it in both places would allow them to upgrade each package independently, instead of having to do it all at once.

**Comment by duckinator** [comment]
Ah, okay. I'm wrapping up for the day, so I think figuring out how to map it to v1 will probably be what I start with tomorrow.

**Comment by mistydemeo** [comment]
I suspect that this probably wants to go into the `BuildLayer` section of config, which is defined here: https://github.com/axodotdev/cargo-dist/blob/455b8eea095c5607d2c44d93d13e52d7de073ce5/cargo-dist/src/config/v1/builds/mod.rs#L49-L75

That file contains non-language-specific build settings, including system dependency requests, so it feels like a natural place to specify the glibc version.

We construct it here: https://github.com/axodotdev/cargo-dist/blob/b00acfeeddfc40e070000d93abe76b6b02d61f55/cargo-dist/src/config/v0_to_v1.rs#L126-L137

We'll also want to make sure `needs_build_layer` also gets extended with the new field to check for.

**Comment by duckinator** [comment]
As best I can tell, the `minimum-glibc-version` option currently starts in the config file and makes it into a `cargo_dist::config::v1::WorkspaceConfig` struct via:

```
config file
-> cargo_dist::config::v0::DistMetadata (.minimum_glibc_version)
-> cargo_dist::config::v1::builds::BuildLayer (.minimum_glibc_version)
-> cargo_dist::config::v1::TomlLayer (.builds.unwrap().minimum_glibc_version)
-> ... (not dug into this yet)
-> cargo_dist::tasks::DistGraphBuilder.new()
-> cargo_dist::config::v1.workspace_config(&WorkspaceGraph, TomlLayer) -> WorkspaceConfig
-> cargo_dist::config::v1::WorkspaceConfig (.builds.minimum_glibc_version)
```


Eventually, `PlatformSupport::new` is called with `DistGraphBuilder`, and `DistGraphBuilder` has a `workspaces: &mut WorkspaceGraph` field (which eventually contains `minimum_glibc_version`). This ultimately calls `native_glibc_version`, but the the `WorkspaceGraph` gets dropped:

```
-> PlatformSupport::new(&DistGraphBuilder, ReleaseIdx)
-> native_runtime_conditions_for_artifact(&DistGraphBuilder, &ArtifactId)
-> native_runtime_conditions_for_asset(&DistManifest, &Option<AssetId>)
-> native_glibc_version(&SystemInfo, &Linkage)
```

_IN THEORY_ I think I just need to bridge that gap between `native_runtime_conditions_for_artifact(&DistGraphBuilder, &ArtifactId)` having access and `native_runtime_conditions_for_asset(&DistManifest, &Option<AssetId>)` not having access.

**Comment by duckinator** [comment]
I have an extremely crude implementation: https://github.com/axodotdev/cargo-dist/pull/1496/commits/fa853e94df18d4e40cf87191cb617a235ba1593a

That is 100% NOT how it should be implemented, but it proves the data is available in `native_runtime_conditions_for_artifact()`.

If anyone has ideas for how to implement it _well_, please let me know!

**Comment by duckinator** [comment]
Demo of the crude implementation working:

image

**Review by mistydemeo — APPROVED** [review]
Looking good. Gave it a test with a valid glibc version for Linux builds where we couldn't figure out the glibc version, and confirmed it's getting piped through.

How about we add this to one of our integration tests? You can pick one of the tests in [this file](https://github.com/axodotdev/cargo-dist/blob/main/cargo-dist/tests/integration-tests.rs) that has `linux-gnu` in the targets and the `shell` installer.

Tested with an invalid `minimum-glibc-version`, and getting the right kinda error:

```
  × failed to parse TOML
  ╰─▶ TOML parse error at line 46, column 25
         |
      46 | minimum-glibc-version = "1.20.1"
         |                         ^^^^^^^^
      libc version must be {major}.{series} where major and series are numbers

    ╭─[/Users/mistydemeo/github/cargodisttest/dist-workspace.toml:46:25]
 45 │ macos-sign = true
 46 │ minimum-glibc-version = "1.20.1"
    ·                         ────────
 47 │
    ╰────
```

**Comment by duckinator** [comment]
Rebased off `main`.

I'm wrapping up for the day right now, but current state is:

I'm getting kinda stuck on adding an integration test for the `minimum-glibc-version` PR. the way I'm testing _manually_ is building a basic project then running `grep 'check_glibc ' target/distrib/${PROJECT_NAME}-installer.sh`.

I can't figure out how to get that information from an integration test.

**Comment by mistydemeo** [comment]
Oh, this is something I should have introduced you to! Our integration tests use snapshotting; generated content like the installers gets written to a snapshot file. One of the things the tests check for is "the snapshot is identical to the content generated during the test". If you make a change that alters how the installers are generated, then that will get tested. Since the glibc content makes it into the shell installers, that means we can safely use the snapshots to test that it worked.

**Comment by duckinator** [comment]
Updated a test to use `minimum-glibc-version` as @mistydemeo suggested, and rebased off master.

This is ready to go as soon as CI passes. :slightly_smiling_face:
